### PR TITLE
Add capability to process AHI's brightness temperature in HSD and NC, cloud mask and cloud property

### PR DESCRIPTION
**Pull request by ibanos90**
## Description

This PR adds the capability to do thinning and/or superobbing on the MPAS mesh for the AHI sensor from Himawari-8 and -9 satellites. For this purpose the `mod_himawari_ahi` module is added and the logic used to process ABI is followed. The processing of  HSD (Himawari Standard Data) brightness temperature (BT) data is based on obs2ioda `hsd.f90` module by Jamie Bresch, which works for one single segment or multiple segments. The processing of NetCDF BT 
data (containing all bands) is added reading data downloaded from the [ftp.ptree.jaxa.jp](url) (registration required). For the cloud mask information (from the [AWS NOAA](https://noaa-himawari9.s3.amazonaws.com/index.html) cloud), this PR adds the capability to read different file names corresponding with changes throughout the years, allowing to process BT and cloud mask whenever available. Caveat, the size of the array is hardcoded `nband+1` in order to process BT and cloud mask together and write them out to the same file. These two products need to be at the same resolution. Other caveats for AHI processing are included in the `mod_himawari_ahi` module.

To enable these changes, the original obs2model taylored for the processing of GOES-ABI data was made more generic, moving some subroutines out of the `mod_goes_abi` module and put in a `utils_mod` module. Some parameters are also moved to the `set_var` module. A new subroutine is added to read the namelist file for both ABI and AHI and an option is added via command line to select what data to process `do_abi` or `do_abi`. In the future it could be added doing both at the same time if needed.

The test folder is refactored to include a test for AHI with corresponding namelist and flist.txt files. Test files are provided to process HSD BT data from Himawari-8 and -9 with corresponding cloud mask as well as NetCDF BT data and CLP (Cloud Property) data. Note that the CLP product is [day-time only](https://www.eorc.jaxa.jp/ptree/userguide.html#302) , see [cldtype_AHI_daytime.pdf](https://github.com/user-attachments/files/18115731/cldtype_AHI_daytime.pdf). A simple python script is provided to plot the `brightness_temperature_13@ObsValue` and `cloudAmount@MetaData` from the superobbed observations in the IODAv1 file.

The name of the executable is also updated along with the `README` to account for these changes.

## Issue closed
None

## List of modified files
M       CMakeLists.txt
M       README.md
R050    goes2mpas/CMakeLists.txt        src/CMakeLists.txt
R063    goes2mpas/goes_abi_mod.F90      src/goes_abi_mod.F90
A       src/himawari_ahi_mod.F90
R086    goes2mpas/main.F90      src/main.F90
R098    goes2mpas/netcdf_mod.f90        src/netcdf_mod.f90
R100    goes2mpas/read_write_indx_mod.F90       src/read_write_indx_mod.F90
R086    goes2mpas/read_write_mpas_mod.F90       src/read_write_mpas_mod.F90
R083    goes2mpas/setvar.F90    src/setvar.F90
A       src/utils_mod.F90
R100    test_abi_read/flist.txt test/abi/flist.txt
R100    test_abi_read/flist.txt_for10kmMPASformat       test/abi/flist.txt_for10kmMPASformat
R100    test_abi_read/flist.txt_forIODAformat   test/abi/flist.txt_forIODAformat
R100    test_abi_read/flist.txt_forMPASformat   test/abi/flist.txt_forMPASformat
R059    test_abi_read/namelist.obs2model        test/abi/namelist.obs2model
R059    test_abi_read/namelist.obs2model_for10kmMPASformat      test/abi/namelist.obs2model_for10kmMPASformat
R069    test_abi_read/namelist.obs2model_forIODAformat  test/abi/namelist.obs2model_forIODAformat
R069    test_abi_read/namelist.obs2model_forMPASformat  test/abi/namelist.obs2model_forMPASformat
R085    test_abi_read/qsub.csh  test/abi/qsub.csh
A       test/ahi/flist.txt
A       test/ahi/flist.txt_H08
A       test/ahi/flist.txt_H08_NC
A       test/ahi/flist.txt_H09
A       test/ahi/flist.txt_for10kmMPASformat
A       test/ahi/flist.txt_forIODAformat
A       test/ahi/flist.txt_forMPASformat
A       test/ahi/namelist.obs2model
A       test/ahi/namelist.obs2model_H08
A       test/ahi/namelist.obs2model_H08_NC
A       test/ahi/namelist.obs2model_H09
A       test/ahi/namelist.obs2model_for10kmMPASformat
A       test/ahi/namelist.obs2model_forIODAformat
A       test/ahi/namelist.obs2model_forMPASformat
A       test/ahi/plot.py
A       test/ahi/qsub.csh
D       test_abi_read/goes2mpas.x

## Tests
Tests conducted on Derecho with GNU compilers. Example of the superobervations for Himawari-9 at 0000Z 1 June 2024: 

bt_cldamount_H09_20240601_00

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -11,30 +11,32 @@ Aim
...
 ```
 source mpas-jedi environment
 git clone https://github.com/byoung-joo/obs2model.git
 cd obs2model
 mkdir build; cd build
 cmake ../ ; make -j4
-cd ../test_abi_read
-./goes2mpas.x
+cd ../test/abi
+ln ../../build/obs2mpas.x .
```
Comment (byoung-joo) [inline README.md]:
```suggestion
ln ../../build/bin/obs2mpas.x .
```

File: test/ahi/flist.txt_for10kmMPASformat
Comment (byoung-joo) [inline test/ahi/flist.txt_for10kmMPASformat:1]: This file can be removed.

File: test/ahi/flist.txt_forIODAformat
Comment (byoung-joo) [inline test/ahi/flist.txt_forIODAformat:1]: This file can be removed.

File: test/ahi/flist.txt_forMPASformat
Comment (byoung-joo) [inline test/ahi/flist.txt_forMPASformat:1]: This file can be removed.

File: test/ahi/namelist.obs2model_for10kmMPASformat
Comment (byoung-joo) [inline test/ahi/namelist.obs2model_for10kmMPASformat:1]: This file can be removed.

File: test/ahi/namelist.obs2model_forIODAformat
Comment (byoung-joo) [inline test/ahi/namelist.obs2model_forIODAformat:1]: This file can be removed.

File: test/ahi/namelist.obs2model_forMPASformat
Comment (byoung-joo) [inline test/ahi/namelist.obs2model_forMPASformat:1]: This file can be removed.

File: src/himawari_ahi_mod.F90 [hunk 2]
```diff
@@ -0,0 +1,1666 @@
...
+               do i = 1, npixel
+                  rdata(it)%rad(ib, i, j) = missing_r
+                  rdata(it)%bt(ib, i, j) = brit(i, j, ib)
+                  rdata(it)%qf(ib, i, j) = missing_r
+                  rdata(it)%sd(ib) = missing_r
+                  F_out(i, j, ib) = brit(i, j, ib)
+
+               end do
+            end do
+            write(varname_out(ib),"(A,I2.2)") 'BT_'//fsat_id//'C', ib
```
Comment (byoung-joo) [inline src/himawari_ahi_mod.F90]: `ib` ranges from `1` to `nband` here, but the typical channel ranges from `7` to `16`. 

Can we use `fband_id` variable here? or we can even set `ib+6` with some assumption. The latter approach actually is being used in `subroutine output_iodav1` and `subroutine output_iodav1_o2m`.

**Review by byoung-joo — COMMENTED** [review]
Thanks for the enhancement, @ibanos90 . Please see my comments.

File: src/himawari_ahi_mod.F90 (hunk 2 above)
Comment (ibanos90) [reply src/himawari_ahi_mod.F90]: This is a good point! I modified it in my latest commit. Thanks @byoung-joo!

File: README.md (hunk 1 above)
Comment (ibanos90) [reply README.md]: Fixed

File: test/ahi/flist.txt_for10kmMPASformat
Comment (ibanos90) [reply test/ahi/flist.txt_for10kmMPASformat:1]: Removed

File: test/ahi/flist.txt_forIODAformat
Comment (ibanos90) [reply test/ahi/flist.txt_forIODAformat:1]: Removed

File: test/ahi/flist.txt_forMPASformat
Comment (ibanos90) [reply test/ahi/flist.txt_forMPASformat:1]: Removed

File: test/ahi/namelist.obs2model_for10kmMPASformat
Comment (ibanos90) [reply test/ahi/namelist.obs2model_for10kmMPASformat:1]: Removed

File: test/ahi/namelist.obs2model_forIODAformat
Comment (ibanos90) [reply test/ahi/namelist.obs2model_forIODAformat:1]: Removed

File: test/ahi/namelist.obs2model_forMPASformat
Comment (ibanos90) [reply test/ahi/namelist.obs2model_forMPASformat:1]: Removed

File: test/ahi/flist.txt_H09_S0110
Comment (byoung-joo) [inline test/ahi/flist.txt_H09_S0110:1]: Just wondering if this testcase is intentionally processing specific segment .

Comment (ibanos90) [reply test/ahi/flist.txt_H09_S0110:1]: Yes, I added this quick testcase for processing only one segment (to prove that either way works) as the other tests were for all 10 segment. Should I remove the one for all H09 segments and keep H08 for all segments and H09 for only one segment?

Comment (byoung-joo) [reply test/ahi/flist.txt_H09_S0110:1]: Thank you for the explanation. 

> Should I remove the one for all H09 segments and keep H08 for all segments and H09 for only one segment?

No, that would be fine!

**Review by byoung-joo — APPROVED** [review]
Thanks for adding this new feature with overall enhancement in the code structure, @ibanos90 !

**Comment by ibanos90** [comment]
> Thanks for adding this new feature with overall enhancement in the code structure, @ibanos90 !

Thanks for your review!!

File: test/ahi/flist.txt_H09_S0110
Comment (ibanos90) [reply test/ahi/flist.txt_H09_S0110:1]: Keeping both for now. Thanks!

**Pull request merged**